azure/services/loadbalancers/spec.go

```diff
@@ -0,0 +1,313 @@
+		frontendIPConfigs = *existingLB.FrontendIPConfigurations
+		wantedIPs, wantedFrontendIDs := getFrontendIPConfigs(*s)
+		for _, ip := range wantedIPs {
+			if !ipExists(frontendIPConfigs, ip) {
+				update = true
+				frontendIPConfigs = append(frontendIPConfigs, ip)
+			}
+		}
+
+		loadBalancingRules = *existingLB.LoadBalancingRules
+		for _, rule := range getLoadBalancingRules(*s, wantedFrontendIDs) {
+			if !lbRuleExists(loadBalancingRules, rule) {
+				update = true
+				loadBalancingRules = append(loadBalancingRules, rule)
+			}
+		}
+
+		backendAddressPools = *existingLB.BackendAddressPools
+		for _, pool := range getBackendAddressPools(*s) {
+			if !poolExists(backendAddressPools, pool) {
+				update = true
+				backendAddressPools = append(backendAddressPools, pool)
+			}
+		}
+
+		outboundRules = *existingLB.OutboundRules
+		for _, rule := range getOutboundRules(*s, wantedFrontendIDs) {
+			if !outboundRuleExists(outboundRules, rule) {
+				update = true
+				outboundRules = append(outboundRules, rule)
+			}
+		}
+
+		probes = *existingLB.Probes
+		for _, probe := range getProbes(*s) {
+			if !probeExists(probes, probe) {
+				update = true
+				probes = append(probes, probe)
+			}
+		}
+
+		if !update {
+			// load balancer already exists with all required defaults
+			return nil, nil
+		}
+	} else {
+		frontendIPConfigs, frontendIDs = getFrontendIPConfigs(*s)
+		loadBalancingRules = getLoadBalancingRules(*s, frontendIDs)
+		backendAddressPools = getBackendAddressPools(*s)
+		outboundRules = getOutboundRules(*s, frontendIDs)
+		probes = getProbes(*s)
+	}
+
+	lb := network.LoadBalancer{
+		Etag:     etag,
+		Sku:      &network.LoadBalancerSku{Name: converters.SKUtoSDK(s.SKU)},
+		Location: to.StringPtr(s.Location),
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: s.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Role:        to.StringPtr(s.Role),
+			Additional:  s.AdditionalTags,
+		})),
+		LoadBalancerPropertiesFormat: &network.LoadBalancerPropertiesFormat{
+			FrontendIPConfigurations: &frontendIPConfigs,
+			BackendAddressPools:      &backendAddressPools,
+			OutboundRules:            &outboundRules,
+			Probes:                   &probes,
+			LoadBalancingRules:       &loadBalancingRules,
+		},
+	}
+
+	return lb, nil
+}
+
+func getFrontendIPConfigs(lbSpec LBSpec) ([]network.FrontendIPConfiguration, []network.SubResource) {
+	frontendIPConfigurations := make([]network.FrontendIPConfiguration, 0)
+	frontendIDs := make([]network.SubResource, 0)
+	for _, ipConfig := range lbSpec.FrontendIPConfigs {
+		var properties network.FrontendIPConfigurationPropertiesFormat
+		if lbSpec.Type == infrav1.Internal {
+			properties = network.FrontendIPConfigurationPropertiesFormat{
+				PrivateIPAllocationMethod: network.IPAllocationMethodStatic,
+				Subnet: &network.Subnet{
+					// TODO: is the VNet resource group the same as the cluster resource group?
```

Comment:
vnet RG is not always the same as the cluster RG since you can BYO pre-existing vnet. Can this TODO be removed?